storage-proofs-update/src/gadgets.rs

```diff
@@ -0,0 +1,467 @@
+use bellperson::{
+    gadgets::{
+        boolean::{field_into_allocated_bits_le, AllocatedBit, Boolean},
+        multipack::pack_bits,
+        num::AllocatedNum,
+    },
+    ConstraintSystem, LinearCombination, SynthesisError,
+};
+use blstrs::Scalar as Fr;
+use ff::{Field, PrimeField};
+use filecoin_hashers::{poseidon::PoseidonHasher, HashFunction, Hasher};
+use storage_proofs_core::gadgets::por::por_no_challenge_input;
+
+use crate::constants::{TreeD, TreeDHasher};
+
+// Allocates `num` as `Fr::NUM_BITS` number of bits.
+pub fn allocated_num_to_allocated_bits<CS: ConstraintSystem<Fr>>(
+    mut cs: CS,
+    num: &AllocatedNum<Fr>,
+) -> Result<Vec<AllocatedBit>, SynthesisError> {
+    let bits = field_into_allocated_bits_le(&mut cs, num.get_value())?;
+    assert_eq!(bits.len(), Fr::NUM_BITS as usize);
+
+    // Assert `(2^0 * bits[0] + ... + 2^(n - 1) * bits[n]) * 1 == num`.
+    let mut lc = LinearCombination::<Fr>::zero();
+    let mut pow2 = Fr::one();
+    for bit in bits.iter() {
+        lc = lc + (pow2, bit.get_variable());
+        pow2 = pow2.double();
+    }
+    cs.enforce(
+        || "binary decomp",
+        |_| lc,
+        |lc| lc + CS::one(),
+        |lc| lc + num.get_variable(),
+    );
+
+    Ok(bits)
+}
+
+pub fn apex_por<CS: ConstraintSystem<Fr>>(
+    mut cs: CS,
+    apex_leafs: Vec<AllocatedNum<Fr>>,
+    partition_bits: Vec<AllocatedBit>,
+    partition_path: Vec<Vec<AllocatedNum<Fr>>>,
+    comm_d: AllocatedNum<Fr>,
+) -> Result<(), SynthesisError> {
+    // `apex_leafs.len()` is guaranteed to be a power of two.
```

Comment:
It is guaranteed because `constants::apex_leaf_count()` is always a power of two and because `apex_por` is not a public gadget; the circuit has already checked (before calling this gadget) that `apex_leafs.len() == apex_leaf_count(sector_size)`.

Reply:
I see